toplevel/genprintval.ml

```diff
@@ -205,34 +206,73 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
        it comes from. Attempt to omit the prefix if the type comes from
        a module that has been opened. *)
 
-    let tree_of_qualified find env ty_path name =
-      match ty_path with
-      | Pident _ ->
-          Oide_ident name
-      | Pdot(p, _s) ->
-          if
-            match get_desc (find (Lident (Out_name.print name)) env) with
-            | Tconstr(ty_path', _, _) -> Path.same ty_path ty_path'
-            | _ -> false
-            | exception Not_found -> false
-          then Oide_ident name
-          else Oide_dot (Out_type.tree_of_path p, Out_name.print name)
-      | Papply _ ->
-          Out_type.tree_of_path ty_path
-      | Pextra_ty _ ->
-          (* These can only appear directly inside of the associated
-             constructor so we can just drop the prefix *)
-          Oide_ident name
+    let tree_of_qualified lookup_all get_path env ty_path name =
+      (* If [ty_path] is [M.N.t] and [name] is [Foo], we want to find
+         a short name for [M.N.Foo] in the current typing environment.
+         Our strategy is to try [Foo], [N.Foo] and [M.N.Foo] in
+         turn. *)
+
+      (* Start by transforming the path [M.N.t] into the Longident [M.N.Foo]. *)
+      let rec lid_of_path on_last =
+        let lid_non_last lid = lid_of_path Fun.id lid in
+        function
+        | Pident last -> Lident (on_last (Ident.name last))
+        | Pdot (p, last) -> Ldot (lid_non_last p, on_last last)
+        | Papply (f, p) -> Lapply (lid_non_last f, lid_non_last p)
+        | Pextra_ty (p, _) ->
+            (* These can only appear directly inside of the associated
+               constructor so we can just drop the prefix *)
+            lid_of_path on_last p
+      in
+      let lid = lid_of_path (fun _tyname -> name) ty_path in
+
+      (* [candidates exn M.N.Foo] is [Foo; N.Foo; M.N.Foo].
+         @raise [exn] on functor application. *)
+      let candidates apply_exn lid =
+        (* [loop M.N [Foo]] is [[Foo]; [N; Foo]; [M; N; Foo]] *)
+        let rec loop lid suff = match lid with
+          | Lident last -> [suff; (last :: suff)]
+          | Ldot(p, s) -> suff :: loop p (s :: suff)
+          | Lapply _ -> raise apply_exn
+        in
+        loop lid [] (* [[]; [Foo]; [N; Foo]; [M; N; Foo]] *)
+        |> List.filter_map Longident.unflatten
+      in
```

Comment:
This implementation feels quite awkward when the environment for constructors and labels contains the list od successive opens. I have the impression that adding a `relative_constructor_longident: Path.t -> t -> Longident`  to `Env` would be more natural.

Reply:
I don't understand the code in env.ml. Maybe TycompTbl is supposed to serve to do this, who knows? I cannot propose an alternative implementation along the line you propose in an acceptable amount of effort, so if you think this is a requirement to implement this feature I will have to close the current PR and let someone else work on it.

Reply:
I will have a look at implementing my suggestion at the top of this PR to make the discussion clearer.

Reply:
@Octachron proposed an implementation of his suggestion in https://github.com/gasche/ocaml/pull/15 , but my understanding from the PR and a brief discussion this morning is that this alternative approach in fact does not really work in presence of shadowing, and it may be that the approach in the present PR is in fact the simplest.